input/postgres/explain_analyze.go

```diff
@@ -0,0 +1,78 @@
+package postgres
+
+import (
+	"context"
+	"database/sql"
+	"fmt"
+	"strings"
+
+	"github.com/guregu/null"
+	"github.com/lib/pq"
+	"github.com/pganalyze/collector/util"
+)
+
+func runExplainAnalyzeWithHelper(ctx context.Context, db *sql.DB, query string, parameters []null.String, parameterTypes []string, analyzeFlags []string, marker string) (explainOutput string, err error) {
+	tx, err := db.BeginTx(ctx, &sql.TxOptions{ReadOnly: true})
+	if err != nil {
+		return "", err
+	}
+	defer tx.Rollback()
+
+	err = tx.QueryRowContext(ctx, marker+"SELECT pganalyze.explain_analyze($1, $2, $3, $4)", marker+query, pq.Array(parameters), pq.Array(parameterTypes), pq.Array(analyzeFlags)).Scan(&explainOutput)
+
+	return
+}
+
+func validateQuery(query string) error {
+	var isUtil []bool
+	// To be on the safe side never EXPLAIN a statement that can't be parsed,
+	// or multiple statements in one (leading to accidental execution)
+	isUtil, err := util.IsUtilityStmt(query)
+	if err != nil || len(isUtil) != 1 || isUtil[0] {
+		err = fmt.Errorf("query is not permitted to run (multi-statement or utility command?)")
+		return err
+	}
+
+	// TODO: Consider adding additional checks here (e.g. blocking known bad function calls)
+
+	return nil
+}
+
+func RunExplainAnalyzeForQueryRun(ctx context.Context, db *sql.DB, query string, parameters []null.String, parameterTypes []string, marker string) (_result string, _err error) {
+	err := validateQuery(query)
+	if err != nil {
+		return "", err
+	}
+
+	// Warm up caches without collecting timing info (slightly faster)
+	_, err = runExplainAnalyzeWithHelper(ctx, db, query, parameters, parameterTypes, []string{"ANALYZE", "TIMING OFF"}, marker)
+
+	if err != nil && !strings.Contains(err.Error(), "statement timeout") {
+		// Run again if it was a timeout error, to make sure we got the caches warmed up all the way
+		_, err = runExplainAnalyzeWithHelper(ctx, db, query, parameters, parameterTypes, []string{"ANALYZE", "TIMING OFF"}, marker)
+		if ctx.Err() != nil {
+			return "", err
```

Comment:
Why is this checking `ctx.Err()` instead of `err`, like all the other cases?

Reply:
Since its only trying to exit early if the context was canceled (e.g. collector shutting down, etc), otherwise we want to retry and ignore the error.

Reply:
Actually, I realized this is unnecessary since we return early on non-timeout errors anyway - will drop this.